tawazi/_dag/constructor.py

```diff
@@ -0,0 +1,121 @@
+
+from .dag import DAG, AsyncDAG
+
+
+def get_args_and_default_args(func: Callable[..., Any]) -> Tuple[List[str], Dict[str, Any]]:
+    """Retrieves the arguments names and the default arguments of a function.
+
+    Args:
+        func: the target function
+
+    Returns:
+        A Tuple containing a List of argument names of non default arguments,
+         and the mapping between the arguments and their default value for default arguments
+
+    >>> def f(a1, a2, *args, d1=123, d2=None): pass
+    >>> get_args_and_default_args(f)
+    (['a1', 'a2', 'args'], {'d1': 123, 'd2': None})
+    """
+    signature = inspect.signature(func)
+    args = []
+    default_args = {}
+    for k, v in signature.parameters.items():
+        if v.default is not inspect.Parameter.empty:
+            default_args[k] = v.default
+        else:
+            args.append(k)
+
+    return args, default_args
+
+
+def __make_dag(
+    _func: Callable[P, RVDAG], max_concurrency: int, is_async: bool
+) -> Union[DAG[P, RVDAG], AsyncDAG[P, RVDAG]]:
+    # 2. make ExecNodes corresponding to the arguments of the ExecNode
+    # 2.1 get the names of the arguments and the default values
+    func_args, func_default_args = get_args_and_default_args(_func)
+
+    # 2.2 Construct non default arguments.
+    # Corresponding values must be provided during usage
+    args: List[ExecNode] = [
+        ArgExecNode(make_axn_id(_func.__qualname__, arg_name)) for arg_name in func_args
+    ]
+    # 2.2 Construct Default arguments.
+    for arg_name, arg in func_default_args.items():
+        axn = ArgExecNode(make_axn_id(_func.__qualname__, arg_name))
+        args.append(axn)
+        node.results[axn.id] = arg
+
+    # 2.3 Arguments are also ExecNodes that get executed inside the scheduler
+    node.exec_nodes.update({exec_node.id: exec_node for exec_node in args})
+    # 2.4 make UsageExecNodes for input arguments
+    uxn_args = [UsageExecNode(exec_node.id) for exec_node in args]
+
+    # 3. Execute the dependency describer function
+    # NOTE: Only ordered parameters are supported at the moment!
+    #  No **kwargs!! Only positional Arguments
+    # used to be fetch the results at the end of the computation
+    returned_val: Any = _func(*uxn_args)  # type: ignore[arg-type]
+
+    returned_usage_exec_nodes: ReturnUXNsType = wrap_in_uxns(_func, returned_val)
+
+    # 4. Construct the DAG/AsyncDAG instance
+    if is_async:
+        return AsyncDAG(
+            qualname=_func.__qualname__,
+            results=node.results,
+            exec_nodes=node.exec_nodes,
+            input_uxns=uxn_args,
+            return_uxns=returned_usage_exec_nodes,
+            max_concurrency=max_concurrency,
+        )
+    return DAG(
+        qualname=_func.__qualname__,
+        results=node.results,
+        exec_nodes=node.exec_nodes,
+        input_uxns=uxn_args,
+        return_uxns=returned_usage_exec_nodes,
+        max_concurrency=max_concurrency,
+    )
+
+
+def _make_dag(
```

Comment:
I'd really rather have a different name for those two functions, it doesn't serve any purpose imho and make the review harder hh

Reply:
done